src/network/networkd-dhcp4.c

```diff
@@ -575,6 +579,144 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
         return 0;
 }
 
+
+static int dhcp4_routing_policy_rule_handler(sd_netlink *rtnl, sd_netlink_message *m, Link *link) {
+        int r;
+
+        assert(link);
+        assert(link->dhcp4_messages > 0);
+
+        link->dhcp4_messages--;
+
+        if (IN_SET(link->state, LINK_STATE_FAILED, LINK_STATE_LINGER))
+                return 1;
+
+        r = sd_netlink_message_get_errno(m);
+        if (r < 0 && r != -EEXIST) {
+                log_link_message_warning_errno(link, m, r, "Could not set DHCPv4 routing policy rule");
+                link_enter_failed(link);
+                return 1;
+        }
+
+        if (link->dhcp4_messages == 0 && !link->network->dhcp_send_decline)
+                dhcp4_check_ready(link);
+        return 1;
+}
+
+static int dhcp4_add_source_routing_policy_rule(Link *link) {
+        struct in_addr address;
+        _cleanup_(routing_policy_rule_freep) RoutingPolicyRule *rule = NULL;
+        RoutingPolicyRule *rrule = NULL;
+        _cleanup_free_ char *from_addr_str = NULL;
+        int r;
+
+        assert(link);
+
+        if (!link->dhcp_lease || !link->network) /* link went down while we configured the IP addresses? */
+                return 0;
+
+        if (!link->network->dhcp_source_routing_enabled)
+                return 0;
+
+        r = sd_dhcp_lease_get_address(link->dhcp_lease, &address);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "DHCP error: no address");
+
+        r = routing_policy_rule_new(&rule);
+        if (r < 0)
+                return log_oom();
+
+        rule->from.in = address;
+        rule->from_prefixlen = 32;
+        rule->family = AF_INET;
+        rule->priority = link->network->dhcp_source_routing_rule_priority;
+        if (link->network->dhcp_route_table_set)
+                rule->table = link->network->dhcp_route_table;
+
+        if (DEBUG_LOGGING) {
+                const union in_addr_union address_union = { .in = address };
+                (void) in_addr_to_string(AF_INET, &address_union, &from_addr_str);
+        }
+
+        log_link_debug(
+                link, "Adding source routing rule %s/%u -> 0.0.0.0/0 table %d priority %d family %d",
+                from_addr_str,
+                rule->from_prefixlen,
+                rule->table,
+                rule->priority,
+                rule->family
+        );
```

Comment:
the `if` block checking for `DEBUG_LOGGING` should cover this `log_link_debug()` call too